pkg/tfcli/init.go

```diff
@@ -0,0 +1,156 @@
+	initLockExists := false
+	err := c.closeOnError(ctx, func() error {
+		var err error
+		initLockExists, err = c.initConfiguration(types.OperationInit, true)
+		if (err == nil || errors.Is(err, cliErrors.OperationInProgressError{})) && initLockExists {
+			if err == nil || cliErrors.IsOperationInProgress(err, types.OperationInit) {
+				return c.removeStateStore()
+			}
+			return nil
+		}
+		return err
+	})
+	if err != nil {
+		return err
+	}
+	// TODO(aru): what if Terraform CLI has crashed before having a chance to
+	// remove the lock?
+	if !initLockExists {
+		// then we need to call an init
+		// TODO(aru): Shared gRPC server configuration will not involve an init lock.
+		return multierr.Combine(c.syncPipeline(ctx, false, pathTerraform, "init", "-input=false"),
+			c.removeStateStore())
+	}
+	return nil
+}
+
+// initConfiguration checks and initializes a Terraform workspace with a proper
+// configuration. If client's workspace does not yet exist, it can prepare
+// workspace dir if mkWorkspace is set.
+// Returns true if Terraform Init lock exists.
+func (c *client) initConfiguration(opType types.OperationType, mkWorkspace bool) (bool, error) {
+	handle, err := c.getHandle()
+	if err != nil {
+		return false, errors.Wrap(err, errInitWorkspace)
+	}
+	c.wsPath = filepath.Join(os.TempDir(), prefixWSDir+handle)
+
+	// check if the workspace already exists, i.e. there is an open operation
+	ok, err := pathExists(c.wsPath, true)
+	if err != nil {
+		return false, err
+	}
+	if !ok && !mkWorkspace {
+		return false, errors.Errorf(fmtErrNoWS, c.wsPath)
+	}
+
+	initLockExists := false
+	if ok {
+		initLockExists, err = pathExists(filepath.Join(c.wsPath, fileInitLock), false)
+		if err != nil {
+			return false, err
+		}
+
+		// check the state lock. If state lock exists, do not overwrite config
+		err = c.checkOperation()
+		if !errors.Is(err, os.ErrNotExist) {
+			return initLockExists, err
+		}
+	}
+	// workspace does not exist & make workspace is requested or
+	// no state lock file
+	if err := os.MkdirAll(c.wsPath, 0755); err != nil {
+		return initLockExists, errors.Wrap(err, errInitWorkspace)
+	}
+
+	conf, err := c.generateTFConfiguration()
+	if err != nil {
+		return initLockExists, errors.Wrap(err, errInitWorkspace)
+	}
+	if err := errors.Wrap(ioutil.WriteFile(filepath.Join(c.wsPath, tplMain), conf, 0644), errInitWorkspace); err != nil {
+		return initLockExists, err
+	}
+
+	xpState := xpState{
+		Operation: opType,
+	}
+	buff, err := json.Marshal(xpState)
+	if err != nil {
+		return initLockExists, errors.Wrap(err, errInitWorkspace)
+	}
+	return initLockExists,
+		errors.Wrap(ioutil.WriteFile(filepath.Join(c.wsPath, fileStateLock), buff, 0644), errInitWorkspace)
+}
+
+func (c *client) checkOperation() error {
+	xpStatePath := filepath.Join(c.wsPath, fileStateLock)
+	// Terraform state lock file does not seem to contain operation type
+	buff, err := ioutil.ReadFile(xpStatePath)
```

Comment:
We had a couple of offline discussions around some possible alternatives to relying on filesystem locks (terraforms and now ours) but would love to revisit them without blocking this PR.

Reply:
Yes, agreed. As discussed previously, I think we can experiment with some alternatives to see how reliable they are by testing if we violate any assumptions by the Terraform CLI, and how scalable would those alternatives be in terms of memory, concurrency, etc.